### PR TITLE
refactor(stackable-telemetry)!: Improve stackable-telemetry to make it easier to build the TraceGuard

### PR DESCRIPTION
**Pull request by NickLarsenNZ**
# Description

Part of https://github.com/stackabletech/issues/issues/639.

## Problem

It is cumbersome to configured the various subscribers to be enabled at runtime. Using @soenkeliebau's code example:

```rust
// Instead of building a TraceGuard, a mutable builer needs to be assigned so that...
let mut builder = Tracing::builder()
    .service_name("whoyougonnacall")
    .with_console_output("WYGC_CONSOLE", LevelFilter::INFO);

// ... builder methods can be called conditionally:
if enable_trace_exporter().context(ParseConfigSnafu)? {
    builder = builder.with_otlp_trace_exporter("TEST_OTLP_TRACE", LevelFilter::TRACE);
}
if enable_log_exporter().context(ParseConfigSnafu)? {
    builder = builder.with_otlp_log_exporter("TEST_OTLP_LOG", LevelFilter::TRACE);
}

// Then finally the trace TraceGuard is returned.
let _tracing_guard = builder.build().init().context(InitializeTelemetrySnafu)?;
```

The purpose of the builder is to allow overriding of the the defaults, however in the current state, the function parameters for the subscribers (console and otlp exporters) break that, and it will only get messier as config options are added.

## Solution

Add common and subscriber specific configuration through a Settings builder (suggested by @Techassi). The previous example would then look like:

```rust
let _trace_guard = Tracing::builder()
    .service_name("whoyougonnacall")
    .with_console_output(
        Settings::builder()
            .env_var("WYGC_CONSOLE")
            .default_level(LevelFilter::INFO)
            .enabled(true)
            // 👇 this part is new, and has no analogue in the example from the problem statement above.
            .console_log_settings_builder()
            .log_format(Format::Plain)
            // ☝️
            .build(),
    )
    .with_otlp_trace_exporter(
        Settings::builder()
            .env_var("TEST_OTLP_TRACE")
            .default_level(LevelFilter::TRACE)
            .enabled(enable_trace_exporter().context(ParseConfigSnafu)?)
            .build(),
    )
    .with_otlp_log_exporter(
        Settings::builder()
            .env_var("TEST_OTLP_LOG")
            .default_level(LevelFilter::TRACE)
            .enabled(enable_log_exporter().context(ParseConfigSnafu)?)
            .build(),
    )
    .build();
```

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

### CODE REVIEW

**Comment by soenkeliebau** [comment]
I like it!

**Comment by NickLarsenNZ** [comment]
> I like it!

It is basically done, there is one thing left to refactor (and enable console log formats), but before merging I will document the pub items (and enable warn on missing docs).

**Comment by NickLarsenNZ** [comment]
This PR was going to include the following items:
- complete docs
- tuples for quicker setup
- console json format output as an option
- 
But they were dropped in order to unblock @xeniape for the RollingFileAppender implementation. I am promising to come back and do those things in a separate PR.

File: crates/stackable-telemetry/src/tracing/settings/console_log.rs
Comment (nightkr) [inline crates/stackable-telemetry/src/tracing/settings/console_log.rs:55]: Wouldn't this also be `Build`? At least it seems inconsistent to sometimes use `From` and sometimes `Build`.

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (nightkr) [inline crates/stackable-telemetry/src/tracing/settings/mod.rs:23]: Would a disabled subscriber even exist beyond the builder stage?

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (nightkr) [inline crates/stackable-telemetry/src/tracing/settings/mod.rs:20]: This could be isolated from each impl by defining a

```rust
fn common(&self) -> &Settings;
```

and then providing default impls for everything else that delegates to it:

```suggestion
    /// Access to the [`Settings::environment_variable`] field.
    fn environment_variable(&self) -> &'static str {
      self.common().environment_variable
    }

    /// Access to the [`Settings::default_level`] field.
    fn default_level(&self) -> LevelFilter {
      self.common().default_level
    }
```

**Review by nightkr — COMMENTED** [review]
Not really relevant since it's excluded from this PR, but the tuple variant smells a lot like "`From` magic" that I would rather avoid.

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (Techassi) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs:20]: Agreed, great idea!

Comment (nightkr) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs:20]: Another alternative would be to have each variant provide a `Deref<Target = Settings>`.

Comment (Techassi) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs:20]: Oh yeah, also a nice idea. I will first try out the `common()` approach and then I can look into the `Deref` solution.

Comment (Techassi) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: First iteration in 5de6ed4

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (Techassi) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: No, but we still need access to the `enabled` field, which is part of the common `Settings` struct.

Comment (nightkr) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: I'm not going to cause a fuss over it right now, but it would be nice to see if we can kill it off eventually.

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (Techassi) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: Second iteration in 8f59b15

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (NickLarsenNZ) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: I'm unsure what you mean, can you describe a little more concretely?

The actual Tracing Subscriber/Layer is never created/registered if `enabled` is false.

File: crates/stackable-telemetry/src/tracing/mod.rs
Comment (Techassi) [inline crates/stackable-telemetry/src/tracing/mod.rs:17]: nit: This should be moved down (not an external crate).

File: crates/stackable-telemetry/src/tracing/mod.rs
Comment (Techassi) [inline crates/stackable-telemetry/src/tracing/mod.rs:26]: nit: I think we _currently_ don't need that attribute.

File: crates/stackable-telemetry/src/tracing/mod.rs
Comment (NickLarsenNZ) [reply crates/stackable-telemetry/src/tracing/mod.rs]: Done in afc21636bf3ec0324aabd3dcf20b65dfe38de946

File: crates/stackable-telemetry/src/tracing/mod.rs
Comment (NickLarsenNZ) [reply crates/stackable-telemetry/src/tracing/mod.rs]: Done in afc21636bf3ec0324aabd3dcf20b65dfe38de946

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (nightkr) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: I haven't tested it or thought it through completely, but in short, you could use `bool::then` and let the `with_foo` take an `Option<FooSettings>`. Something like:

```rust
.with_otlp_log_exporter(
  otlp_log_flag.then(|| Settings::builder()
    .blah()
    .build()
  )
)
```

**Review by Techassi — APPROVED** [review]

**Review by Techassi — APPROVED** [review]

File: crates/stackable-telemetry/src/tracing/settings/mod.rs
Comment (NickLarsenNZ) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: I have added an item to the parent issue: https://github.com/stackabletech/issues/issues/639

Comment (NickLarsenNZ) [reply crates/stackable-telemetry/src/tracing/settings/mod.rs]: With the help of @Techassi this has been done in https://github.com/stackabletech/operator-rs/pull/951

**Pull request merged**
